src/Host/Monitor/Impl/CredentialManager.cs

```diff
@@ -0,0 +1,418 @@
+    public class CredentialManager {
+        private const string BrokerUserCredName = "RHostBrokerUserCred";
+
+        public static Task<bool> IsBrokerUserCredentialSavedAsync(ILogger logger = null) {
+            return Task.Run(() => {
+                IntPtr credPtr;
+                bool credStatus = false;
+                if (CredRead(BrokerUserCredName, CredType.GENERIC, 0, out credPtr)) {
+                    credStatus = true;
+                } else {
+                    int error = Marshal.GetLastWin32Error();
+                    if (error == Win32ErrorCodes.ERROR_NOT_FOUND) {
+                        credStatus = false;
+                        logger?.LogDebug(Resources.Info_DidNotFindSavedCredentails);
+                    }
+                }
+                if (credPtr != IntPtr.Zero) {
+                    CredFree(credPtr);
+                }
+                return credStatus;
+            });
+        }
+
+        private const int CRED_MAX_USERNAME_LENGTH = 256;
+        private const int CRED_MAX_DOMAIN_LENGTH = 256;
+        private const int CRED_MAX_PASSWORD_LENGTH = 256;
+
+        public static int MaxCredUIAttempts => 3;
+        public static async Task<int> GetAndSaveCredentialsFromUserAsync(ILogger logger = null) {
+            await TaskUtilities.SwitchToBackgroundThread();
+
+            CredUIInfo credInfo = new CredUIInfo();
+            credInfo.cbSize = Marshal.SizeOf(credInfo);
+            credInfo.pszCaptionText = Resources.Text_RHostBrokerCredentials;
+            credInfo.pszMessageText = Resources.Text_RHostBrokerCredentialsDetail;
+            credInfo.hwndParent = Process.GetCurrentProcess().MainWindowHandle;
+
+            int errorCode = 0;
+            uint authenticationPackage = 0;
+            IntPtr credentialBuffer;
+            uint credentialSize;
+            bool saveCreds = false;
+
+            int attempts = 0;
+            while (attempts++ < MaxCredUIAttempts) {
+                logger?.LogInformation(Resources.Info_CredAttempt, attempts);
+                uint ret = CredUIPromptForWindowsCredentials(ref credInfo, errorCode, ref authenticationPackage, IntPtr.Zero, 0, out credentialBuffer, out credentialSize, ref saveCreds, PromptForWindowsCredentialsFlags.CREDUIWIN_GENERIC);
+
+                // User clicked the cancel button.
+                if (ret == Win32ErrorCodes.ERROR_CANCELLED) {
+                    logger?.LogInformation(Resources.Info_CredUIDismissed);
+                    break;
+                } else if (ret == Win32ErrorCodes.ERROR_SUCCESS) {
+                    // User entered credentials
+                    logger?.LogInformation(Resources.Info_CredUICredsReceived);
+                    StringBuilder username = new StringBuilder(CRED_MAX_USERNAME_LENGTH);
+                    StringBuilder domain = new StringBuilder(CRED_MAX_USERNAME_LENGTH);
+                    StringBuilder password = new StringBuilder(CRED_MAX_USERNAME_LENGTH);
```

Comment:
I created a work item for this. I will user the APIs you created after your code goes in. #2460